app/src/main/java/com/example/androidtraining/MainActivity.kt

```diff
@@ -13,11 +14,17 @@ import androidx.compose.ui.Modifier
 import androidx.navigation.compose.NavHost
 import androidx.navigation.compose.composable
 import androidx.navigation.compose.rememberNavController
+import com.example.androidtraining.networking.TokenProvider
 import com.example.androidtraining.screens.login.LoginScreen
 import com.example.androidtraining.screens.product.ProductScreen
 import com.example.androidtraining.ui.theme.AndroidTrainingTheme
+import dagger.hilt.android.AndroidEntryPoint
+import javax.inject.Inject
 
+@AndroidEntryPoint
 class MainActivity : ComponentActivity() {
+    @Inject
+    lateinit var tokenProvider: TokenProvider
```

Comment:
`TokenProvider` should be injected in a viewModel directly. The idea is that the view doesn't need to know about anything non-view/(business) logic-related stuff. You can safely do that and figure out all the logic in the viewModel. The login should only be notified if there is an error or if there is a successful login.

Reply:
fixed